### PR TITLE
[Inductor] Eliminate redundant to_dtype node

### PR DESCRIPTION
**Pull request by EikanWang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #97070
* __->__ #96650

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @soumith @voznesenskym @penguinwu @anijain2305 @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/codegen/cpp.py [hunk 1]
```diff
@@ -2062,9 +2059,36 @@ def add_to_dtype(sub_graph: torch.fx.Graph):
                     pass
 
             def eliminate_to_dtype(sub_graph: torch.fx.Graph):
-                # TODO(Eikan) Remove redundant to_dtype like load_bf16 + to_fp32 + to_bf16 + store_bf16
-                # => load_bf16 + store_bf16
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:2066]: The new function removes the consecutive to_dtype with the same dtype. Are you also going to cancel the "to_fp32 + to_bf16" as illustrated in the comment above?

File: torch/_inductor/codegen/cpp.py [hunk 2]
```diff
@@ -2062,9 +2059,36 @@ def add_to_dtype(sub_graph: torch.fx.Graph):
...
+                    for node_users in all_to_nodes_and_users:
+                        for node, users in node_users.items():
+                            if all(usr.args[-1] == node.args[-1] for usr in users):
+                                val_node = node.all_input_nodes[-1]
+                                node.replace_all_uses_with(val_node)
+                                sub_graph.erase_node(node)
+
+                    sub_graph.lint()
+
+                def _eliminate_to_node_mem_copy(sub_grah: torch.fx.Graph):
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:2087]: What does this function do?

File: test/inductor/test_torchinductor.py [hunk 3]
```diff
@@ -6296,6 +6296,23 @@ def fn(x):
                         - metrics.generated_cpp_vec_kernel_count
                     ) == 0
 
+        def test_redundant_to_node_elimination_bf16(self):
+            def fn(x, y):
+                res = x + y
```
Comment (jgong5) [inline test/inductor/test_torchinductor.py:6301]: Can you explain how this test would produce redundant to_dtype to float?

File: torch/_inductor/codegen/cpp.py (hunk 1 above)
Comment (EikanWang) [reply torch/_inductor/codegen/cpp.py:2066]: Move this feature to `_eliminate_to_node_mem_copy`.

File: torch/_inductor/codegen/cpp.py (hunk 2 above)
Comment (EikanWang) [reply torch/_inductor/codegen/cpp.py:2087]: `load_bf16 + to_fp32 + to_bf16 + store_bf16` => `load_bf16 + store_bf16`

File: test/inductor/test_torchinductor.py (hunk 3 above)
Comment (EikanWang) [reply test/inductor/test_torchinductor.py:6301]: This test case invokes `mean` and the lowering of `mean` inserts `to_dtype`. https://github.com/pytorch/pytorch/blob/master/torch/_inductor/lowering.py#L3464

**Review by jgong5 — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**